sinks/gcm.go

```diff
@@ -55,112 +69,123 @@ func NewSink() (*GcmSink, error) {
-	for _, pod := range data.Pods {
-		metrics = append(metrics, self.podToMetrics(&pod)...)
-	}
-	metrics = append(metrics, self.rawContainersToMetrics(data.Containers)...)
-	metrics = append(metrics, self.rawContainersToMetrics(data.Machine)...)
-
-	// Push the metrics step size at a time. Try to push all the metrics even if there are errors.
+	// Format metrics and push them.
 	var lastErr error
-	step := self.driver.MaxNumPushMetrics()
-	for i := 0; i < len(metrics); i += step {
-		endIndex := i + step
-		if endIndex > len(metrics) {
-			endIndex = len(metrics)
-		}
-
-		err := self.driver.PushMetrics(metrics[i:endIndex])
+	for _, pod := range data.Pods {
+		err := self.podToMetrics(&pod)
 		if err != nil {
 			lastErr = err
 		}
 	}
+	err := self.rawContainersToMetrics(data.Containers)
+	if err != nil {
+		lastErr = err
+	}
+	err = self.rawContainersToMetrics(data.Machine)
+	if err != nil {
+		lastErr = err
+	}
 
 	return lastErr
 }
 
-// Concatenates a map of labels into a comma-separated key=value pairs.
-func labelsToString(labels map[string]string) string {
-	var buffer bytes.Buffer
-	first := true
-	for key, value := range labels {
-		if !first {
-			buffer.WriteString(",")
-		}
-		first = false
-		buffer.WriteString(key)
-		buffer.WriteString("=")
-		buffer.WriteString(value)
-	}
-	return buffer.String()
-}
-
-func (self *GcmSink) podToMetrics(pod *sources.Pod) []Metric {
-	metrics := make([]Metric, 0, len(pod.Containers))
-
+func (self *GcmSink) podToMetrics(pod *sources.Pod) error {
 	// Generate the labels.
 	labels := make(map[string]string)
 	labels[labelPodId] = pod.ID
-	labels[labelLabels] = labelsToString(pod.Labels)
+	labels[labelLabels] = gcm.LabelsToString(pod.Labels)
 	labels[labelHostname] = pod.Hostname
 
 	// Break the individual metrics from the container statistics.
+	var lastErr error
 	for _, container := range pod.Containers {
-		metrics = append(metrics, self.containerToMetrics(container, labels)...)
+		err := self.containerToMetrics(container, labels)
+		if err != nil {
+			lastErr = err
+		}
 	}
-	return metrics
-}
 
-func (self *GcmSink) rawContainersToMetrics(containers []sources.RawContainer) []Metric {
-	metrics := make([]Metric, 0, len(containers))
+	return lastErr
+}
 
+func (self *GcmSink) rawContainersToMetrics(containers []sources.RawContainer) error {
 	labels := make(map[string]string)
+	var lastErr error
 	for _, container := range containers {
 		labels[labelHostname] = container.Hostname
-		metrics = append(metrics, self.containerToMetrics(&container.Container, labels)...)
+		err := self.containerToMetrics(&container.Container, labels)
+		if err != nil {
+			lastErr = err
+		}
 	}
-	return metrics
+
+	return lastErr
 }
 
-func (self *GcmSink) containerToMetrics(container *sources.Container, labels map[string]string) []Metric {
+func (self *GcmSink) containerToMetrics(container *sources.Container, labels map[string]string) error {
```

Comment:
Since this method pushes the data in addition to decoding it, consider renaming it or splitting the pushing part into a separate method.